Feat: root layout( sidebar ) 컴포넌트 작업

leeleeleeleejun opened the pull request:
## #️⃣연관된 이슈

> close #8 

## 📝작업 내용
- svg아이콘의 fill 등 속성을 변경하기 위해 svgr 라이브러리 설치
 9a8fce079e53c98fdc9611b3466bb405e5132773 8fa52a14c4af9df326031993bc124bcd95935716
- 공통 svg아이콘 추가 및 sidebar에서 사용되는 아이콘 복사 2026b3e46a205cef2296eea12d015c9b41f33161
 ( fill 속성 변경 필요 때문에 중복으로 svg파일 허용함 )
- 로그인 페이지에서 sidebar를 가진 레이아웃을 사용하지 않기 때문에 (with-sidebar)디렉토리 생성 후 해당 디렉토리의 page를 메인(홈)으로 만들었습니다. 9317df5fc6b67b9e0e9b214f5d524f9802377d68
- 사이드바 컴포넌트  및 하위 컴포넌트 생성

### 스크린샷 (선택)
사이드바 레이아웃 사용안하는 페이지 존재
<img width="200" src="https://github.com/user-attachments/assets/88921096-bbdf-4003-97c6-06acb7038bba" />

## 💬리뷰 요구사항(선택)
svg아이콘 속성 변경이 필요해 svgr를 사용해 컴포넌트에서 속성을 지정하려 했지만,
기본으로 svg파일에서 지정되어 있는 값의 우선순위가 높아 적용이 되지 않습니다.
그렇기 때문에 기존에 지정되어 있는 속성의 값(색상)을 공백 `fill = ""` 으로 두어서 사용해야 한다고 판단했습니다.
하지만 재사용 시 불편할 수 있을 것 같아 원래 파일을 변경하지 않고 `assets/icons/sidebar` 디렉토리를 생성해 같은 아이콘을 복사 후 수정하고 있습니다.
파일을 중복하지 않고 다른 좋은 방법이 있다면 피드백 부탁드립니다!

leeleeleeleejun committed:
chore: svgr 라이브러리 설치 및 초기설정

leeleeleeleejun committed:
chore: svgr타입 파일 tsconfig에 추가

leeleeleeleejun committed:
chore: svg icon 추가 및 디렉토리 정리
- sidebar에서 사용되는 svg 아이콘 수정

leeleeleeleejun committed:
chore: (with-sidebar) 디렉토리 생성 및 페이지 이동

leeleeleeleejun committed:
feat: layout sidebar 생성
 - 내부 공통 컴포넌트 sidebarItem 생성

yoonncho commented on src/app/(with-sidebar)/layout.tsx:
```diff
@@ -0,0 +1,13 @@
+  )
+}
+
+export default WithSidebarLayout
```
사이드바 유무로 디렉토리 나누는 방식도 좋아보이네요!

다만 저희 서비스 봤을 때 사실 로그인 페이지 빼고는 다 `SideBar`가 존재하는 것 같더라구요.
앞으로 더 생길수는 있지만 디렉토리가 한 depth 깊어지는게 오히려 구조가 복잡해지는 느낌이 들어서요 🤔
한 곳 뿐이라면 차라리 하나의 layout.tsx에서 
```ts
const HIDE_SIDEBAR_PAGES = [“/login-page]; // 예외적으로 사이드바가 존재하지 않는 페이지
const pathname = usePathname(); 

{!HIDE_SIDEBAR_PAGES.includes(pathname) && <Sidebar />}
```

이런식으로 체크해서 미노출 시키는게 더 간단하지 않을까 싶은데 어떻게 생각하시나요?

leeleeleeleejun replied:
윤서님의 제안은 두 가지 큰 장점이 있다고 생각됩니다

1. 사이드바를 사용하지 않는 페이지들을 코드상에서 한눈에 파악할 수 있어 가독성이 좋습니다
2. 최상위 layout에서 관리되므로 디렉토리 구조의 복잡성을 줄일 수 있습니다

다만, 페이지 이동 시 주소를 비교하는 연산( 조건부 렌더링 로직 ) 추가와 `client component( usePathname )` 가 생성된다는 점을 고려했습니다. 하지만 가벼운 연산이므로 성능에 미치는 영향은 미미할 것으로 판단됩니다.

장단점을 비교했을 때, 윤서님이 제안하신 방식이 더 많은 이점이 있다고 판단되어 해당 방식으로 변경하도록 하겠습니다!

한 가지 고려사항은 최상위 layout은 모든 페이지에서 공유되기 때문에
여기서 `client component( usePathname )`를 사용하면 전체 애플리케이션의 초기 로딩 성능에 영향을 줄 수 있습니다.
따라서 페이지 구분 로직은 sidebar컴포넌트에 위치시키도록 하겠습니다.

많은 것을 생각할 수 있도록 해주신 좋은 피드백 감사합니다! 😊

yoonncho replied:
네 이견없습니당!
페이지 하나이다 보니까 단순한 구조가 좋을 것 같다 생각했는데
서비스의 특징마다 다르겠네요. 잘 정리해주셔서 감사합니다:)

yoonncho replied:
네 이견 없습니당!
페이지 하나이다 보니까 단순한 구조가 나을 것 같다 생각했는데
서비스 특징마다 다르겠군요. 잘 정리해주셔서 감사합니다 :)

yoonncho commented on src/app/(with-sidebar)/_components/SidebarItem/SearchButton.tsx:
```diff
@@ -0,0 +1,22 @@
+  )
+}
+
+export default SearchButton
```
저희 이번에 통합 검색 미포함으로 결정했던 것 같아요!! (사이드바에도 노출x)

leeleeleeleejun replied:
앗 피그마에선 변경사항이 없길래 몰랐습니다
디자이너분들에게 더블체크 후 변경된 디자인 적용하겠습니다!

yoonncho commented on src/app/(with-sidebar)/_components/NavigateList/index.tsx:
```diff
@@ -0,0 +1,48 @@
+import HomeIcon from '@/assets/icons/sidebar/non-active/home.svg'
+import Edit from '@/assets/icons/sidebar/non-active/edit.svg'
+import Profile from '@/assets/icons/sidebar/non-active/profile.svg'
+import ProfileUser from '@/assets/icons/sidebar/non-active/profile-user.svg'
```
아이콘을 불러올 때는 위의 `HomeIcon`과 동일하게 `-Icon`으로 통일하면 어떨까요?
(예: `ProfileIcon`, `ProfileUserIcon`)

현재는 일반 컴포넌트처럼 헷갈려서요!!

leeleeleeleejun replied:
앗 나중에 통일한다는 걸 깜빡했네요 ㅠ
확인감사합니다!

yoonncho commented on src/app/(with-sidebar)/_components/NavigateList/index.tsx:
```diff
@@ -0,0 +1,48 @@
+  return (
+    <nav>
+      <SidebarItem
+        path={'/'}
```
저희 단순 문자열만 넘길때는 중괄호 없이 작성하기로 하지 않았나여?
동적값이 아닌 경우 중괄호가 없어도 될 것 같습니다~

leeleeleeleejun replied:
아 일단 임시로 해놓은 거라 따로 신경쓰지 않았는데 중괄호 없는 것으로 변경해놓겠습니다!
그리고 path 같은 경우에 상수로 관리하면 좋을 거 같은데 어떠신가요??
ex)
`export const CLIENT_PATH = {  HOME: '/', LOGIN: '/login', ... };`
`export const API_PATH = {}`

yoonncho replied:
네 좋아요~ 조만간 각 페이지에 대한 url도 논의해요!

yoonncho commented on svgr.d.ts:
```diff
@@ -0,0 +1,10 @@
+  export default content
+}
+
+declare module '*.svg?url' {
```
svg url로 사용할 경우를 위한 걸까요?

leeleeleeleejun replied:
넵 맞습니다!
https://react-svgr.com/docs/next/ 공식 문서에 나온 사용예시 입니다!
```
import Image from 'next/image'
import starUrl from './star.svg?url'

const Example = () => (
  <div>
    <Image src={starUrl} />
  </div>
)
```

yoonncho commented on src/app/(with-sidebar)/_components/SidebarItem/index.tsx:
```diff
@@ -0,0 +1,46 @@
+  nonActiveIcon,
+  contentName,
+  path,
+  ...props
```
`SidebarItem`이 단순히 해당 `path`로 이동하는 역할이라면 추가적인 속성이 불필요하지 않을까 싶은데요!
예를 들어 `<SidebarItem path="/aa" href="bb" />`로 중복 작성하는 경우 예상하지 않은 방식으로 동작할 것 같아서요.

추후 확장해야할 필요성이 있으면 추가하면 어떨까하는 의견입니당

leeleeleeleejun replied:
피드백 하신대로 타입 안정성을 위해 ...props 문 삭제했습니다~!

yoonncho reviewed (commented):
안그래도 아이콘 세팅하고 있었는데 감사합니다~
저도 이번주 내에 후딱 작업해서 PR 올리겠습니다:)

yoonncho commented:
아이콘 관련해서는, 중복해서 추가하신 이유가 stroke, fill 색상 때문일까요?
`stroke='current'`로 두고 `stroke: white` 이런식으로 적용하면 어떨까요?
fill도 `fill='current'`로 하구요!

yoonncho commented:
사소한거긴 하지만 브랜치명을 `feat/-`, `refactor/-` 요런식으로 통일하면 어떨까요?! 
커밋메시지의 타입과 동일하게요 ㅎㅎ (feat, chore, refactor ..)

leeleeleeleejun committed:
refactor: svg아이콘 import명 -icon 추가 통일

leeleeleeleejun committed:
refactor: sideba-item-props에서 ...props문 삭제 (확장 X)

leeleeleeleejun committed:
refactor: props string만 전달 시 중괄호 제외

leeleeleeleejun committed:
refactor: search-button 컴포넌트 삭제

leeleeleeleejun committed:
refactor: 디렉토리 구조 변경
- (with-sidebar) 디렉토리 삭제
- sidebar컴포넌트 최상위 layout 파일로 이동
- sidebar컴포넌트에서 페이지 이동 시 주소를 비교하는 연산 추가

leeleeleeleejun committed:
chore: storybook svgr사용 설정

leeleeleeleejun committed:
feat: sidebar 컴포넌트 스토리북 추가
- 색상 구분을 위해 임시 url 수정

yoonncho commented on src/components/RootLayout/Sidebar/index.tsx:
```diff
@@ -0,0 +1,31 @@
+  const pathname = usePathname()
+
+  return (
+    !HIDE_SIDEBAR_PAGE_LIST.includes(pathname) && (
```
아예 early return 하면 어떨까요?
사이드바가 미노출되는 조건을 한 눈에 파악할 수 있을 것 같아요!

```ts
if(HIDE_SIDEBAR_PAGE_LIST.includes(pathname)) return null
```

leeleeleeleejun replied:
아 넵! 
코딩 컨벤션에 맞게 early return 적용하겠습니다~

yoonncho commented on src/components/RootLayout/SidebarItem/NotificationLink.tsx:
```diff
@@ -0,0 +1,32 @@
+import { cn } from '@/utils/cn'
+import SidebarItem from './index'
+import Notification from '@/assets/icons/sidebar/non-active/notification.svg'
+import ActiveNotification from '@/assets/icons/sidebar/active/notification-filled.svg'
```
아직 `-Icon` 변경 필요한 부분이 남아있네요!

(아이콘 파일을 매번 중복 추가하는 방법 외에 저도 확인해보겠습니다)

yoonncho commented on src/components/RootLayout/SidebarItem/index.tsx:
```diff
@@ -0,0 +1,44 @@
+import { cn } from '@/utils/cn'
+import { usePathname } from 'next/navigation'
+
+export interface SidebarItemProps extends React.ComponentPropsWithoutRef<'a'> {
```
추가 props를 안 받는다면 요기 인터페이스 확장도 불필요할 것 같습니당

yoonncho reviewed (approved):
고생많으셨습니다 👍👍

leeleeleeleejun committed:
refactor: svg아이콘 import명 -Icon 추가 통일

leeleeleeleejun committed:
refactor: sidebar-item-props 타입에 확장문 제거

leeleeleeleejun committed:
refactor: sidebar 노출 조건 if문 early return 적용

leeleeleeleejun commented:
추가 피드백 모두 적용했습니다~!
